### PR TITLE
Making the tags work

### PR DESCRIPTION
**Pull request by sebbel**
Putting this to work in progress. Apparently some work has already been done with the creation of the 'tagged-courses' routes and templates. Did we not already connect this functionality or is it a stub?

Please take a look at this @Crowducate/development

### CODE REVIEW

**Comment by sebbel** [comment]
We need to discuss on how many tags we want to display.

**Comment by brylie** [comment]
Right, we can use a truncate helper to limit the number of tags displayed in the template.

**Comment by brylie** [comment]
Also, the course category is a different field than tags. It was intended for the Category field to allow a single selection from a pre-determined taxonomy of subjects. This contrasts with the Tags field, which is a free-form tagging interface.

**Comment by sebbel** [comment]
Tags feature is now working, but I am currently debugging why it is not working when I add the code inside the taggedCourse Controller. The subscription and action resides in the routes for now.

File: both/router/routes.js [hunk 1]
```diff
@@ -8,15 +8,26 @@ Router.route('/teach');
 
 Router.route('/learn');
 
+Router.route('/learn/:tag', {
+  subscriptions: function() {
+    return Meteor.subscribe('taggedCourses', this.params.tag)
```
Comment (brylie) [inline both/router/routes.js]: Avoid router level subscriptions. Instead, [add subscriptions at the template level](http://guide.meteor.com/data-loading.html#organizing-subscriptions). Specifically, move this subscription to the `taggedCourses` template `created()` callback.

File: client/templates/shared/course-card.html [hunk 2]
```diff
@@ -7,7 +7,7 @@
                 </a>
             </div>
             <div class="caption">
-                <a class="course-category" href="#">Mathematics</a>
+              <a class="course-category" href="/learn/{{keywords.[0]}}">{{keywords.[0]}}</a>
```
Comment (brylie) [inline client/templates/shared/course-card.html]: The course **category** (or _subject_) and **keyword tags** are intended as separate fields.  For the Course Category feature, you will need to add a new field to the `teach` template, called `category`. The `category` field will offer a pre-defined list of `allowedValues`, such as:
- Science & Technology
- Humanities & Art
- Medicine and Health
- Social Sciences
- Other

Please see the ["Course subjects or topics" design document](https://crowducate.hackpad.com/Course-subjects-or-topics-Z3E7jEYxu1E) for further details.

**Comment by sebbel** [comment]
Thanks for your review @brylie

File: both/router/routes.js (hunk 1 above)
Comment (sebbel) [inline both/router/routes.js]: I don't know if I can still access the params from the template.

File: client/templates/shared/course-card.html (hunk 2 above)
Comment (sebbel) [inline client/templates/shared/course-card.html]: category should be handled in another PR, i think.

**Comment by sebbel** [comment]
I have reverted course category to its former state and added a tag link, similar to categories. It just looks like another category for now.

**Comment by sebbel** [comment]
Hitting a roadblock here. I have moved the subscription to the template

```
Template.taggedCourses.onCreated(function(){
      this.autorun(function(){
        this.subscribe('taggedCourses', this.params.tag)
      });
});
```

But now I've lost functionality, as my route is now showing all published courses, regardless of selected tag. I have not pushed the changes as they are breaking atm.

Route for Tags as follows

```
Router.route('/learn/:tag', function() {
    this.render('taggedCourses')
});
```

I suspect that this.params.tag is not available on the template and then the Publish of Tagged Courses runs into a wildcard on keywords, which is essentially Courses.find();

**Comment by sebbel** [comment]
I can now extract the params from within the Template. Will push changes when I get functionality back.

**Comment by sebbel** [comment]
Working for now. Do I need to keep the subscription wrapped in a `this.autorun` @brylie ?

**Comment by brylie** [comment]
The subscription does not need to be wrapped in `autorun()`.

File: both/router/routes.js [hunk 3]
```diff
@@ -8,15 +8,21 @@ Router.route('/teach');
...
+    this.render('taggedCourses')
+});
+
 Router.route('/profile')
 
 Router.route('/profileSettings');
 
 Router.route('/settings');
 
+/*
```
Comment (brylie) [inline both/router/routes.js]: Clean up this unused route.

File: client/templates/courses/tagged-courses.js [hunk 4]
```diff
@@ -6,3 +6,11 @@ Template.taggedCourses.helpers({
         return Session.get('tag');
     }
 });
+
+Template.taggedCourses.onCreated(function(){
+  var controller = Iron.controller();
```
Comment (brylie) [inline client/templates/courses/tagged-courses.js]: Add comments here, describing in plain language the purpose of each line.

File: client/templates/courses/tagged-courses.js [hunk 5]
```diff
@@ -6,3 +6,11 @@ Template.taggedCourses.helpers({
...
+  var params = controller.getParams();
+  this.autorun(function(){
```
Comment (brylie) [inline client/templates/courses/tagged-courses.js]: Remove `autorun()`. 

The `autorun()` is used for code that should re-run when reactive data sources, such as database cursors (`Collection.find()`) change. Since there is no reactive data source within your `autorun()` function, there is no need for `autorun()` at all.

File: client/templates/shared/course-card.html [hunk 6]
```diff
@@ -7,7 +7,8 @@
                 </a>
             </div>
             <div class="caption">
-                <a class="course-category" href="#">Mathematics</a>
+              <a class="course-category" href="#">Mathematics</a> 
+              <a class="course-tags" href="/learn/{{keywords.[0]}}">{{keywords.[0]}}</a>
```
Comment (brylie) [inline client/templates/shared/course-card.html]: This should be reverted to `course-category`, since [course tags are displayed in the course card footer](https://github.com/Crowducate/crowducate-platform/blob/develop/client/templates/shared/course-card.html#L18-L22).

Comment (brylie) [inline client/templates/shared/course-card.html]: In other words, course category and course tags are separate features.

File: both/router/routes.js (hunk 3 above)
Comment (sebbel) [inline both/router/routes.js]: profile, profilesettings and settings?

**Comment by sebbel** [comment]
I have followed your suggestions and I think we are getting close to an acceptable state.
As we bypass the taggedCourse controller, are we positive to delete it?

**Comment by brylie** [comment]
Which controller? I am not sure that we need to delete any controller. Just clean up the unused route.

**Comment by sebbel** [comment]
The taggedCourses controller.

**Comment by brylie** [comment]
I did not suggest to delete the controller. What are your thoughts?

**Comment by brylie** [comment]
I reviewed the `taggedCourses` controller, and it looks like we only need [one line](https://github.com/Crowducate/crowducate-platform/blob/2f748d7de6a4422ba6b71b49d372f2e79ceaa5ac/both/controllers/taggedCourses.js#L23). Perhaps the rest of the controller can be cleaned up as well.

**Comment by sebbel** [comment]
While testing I have encountered some inconsistencies. could @brylie or @Rahbaran take a look again? 
1. Go to learn
2. Click on a tag
3. You are redirected to /learn/tag
4. If you click on another tag while in this view, tag functionality is not working anymore.

**Comment by Rahbaran** [comment]
I'm going to have a look at the functionality this evening.

**Comment by sebbel** [comment]
Thanks @Rahbaran, I am not sure if the bugs are because of my machine or because of a fault that I did.

**Comment by Rahbaran** [comment]
I think the current status just shows ALL courses that have at least a tag. I have 5 courses, 3 with tags, 2 are without any tags:

clicking on ANY tag yields the same result.

so e.g. `http://localhost:3000/learn/bees`  even shows tags without bees but doesn't show the other two tagless courses

File: client/templates/courses/tagged-courses.js [hunk 7]
```diff
@@ -6,3 +6,12 @@ Template.taggedCourses.helpers({
...
     }
 });
+
+Template.taggedCourses.onCreated(function(){
+  // Accessing the Iron.controller to invoke getParams method of Iron Router.
+  var controller = Iron.controller();
+  // Getting Params of the URL
+  var params = controller.getParams();
+  // Invoking subscription on the Courses with the :tag params
+  Meteor.subscribe('taggedCourses', params.tag)
```
Comment (brylie) [inline client/templates/courses/tagged-courses.js]: Check the logic inside of the [`taggedCourses` publication](https://github.com/Crowducate/crowducate-platform/blob/develop/server/publications/courses.js#L10), to make sure it only returns a cursor with matching tagged courses.

**Comment by brylie** [comment]
It sounds like we may have an issue with the publications that are active when viewing the taggedCourses route. I.e. there may be multiple subscriptions active, even at the global level, and our template logic may not be properly filtering the results. I will take a look.

**Comment by brylie** [comment]
Yep, it looks like all, or most, of our subscriptions are within Controllers. We should start moving those to the template level, inside ['smart components'](http://guide.meteor.com/blaze.html#smart-components) to make our code easier to maintain.

**Comment by brylie** [comment]
The problem was in oversubscribing to courses and not controlling what courses were returned in the `courses` template helper. I have committed fixes and cleaned up the code a bit.

File: both/controllers/taggedCourses.js [hunk 8]
```diff
@@ -1,25 +0,0 @@
-TaggedCoursesController = AppController.extend({
-    waitOn: function() {
-        // Wait for the tagged resources to be available for the specific tag
-        return [
-            this.subscribe('taggedCourses', this.params.tag),
-            this.subscribe('images')
-        ];
-    },
-    data: function () {
-        return Courses.find();
```
Comment (brylie) [inline both/controllers/taggedCourses.js:10]: This line set the data context to include **all courses**, which was why more courses were appearing than the current tag.

File: client/templates/courses/tagged-courses.js [hunk 9]
```diff
@@ -1,8 +1,49 @@
...
     'courses': function () {
-        return Courses.find().fetch();
+      // Get reference to template instance
+      var instance = Template.instance();
+
+      // Get tag from template instance
+      var tag = instance.tag;
+
+      // Fetch courses matching current tag
+      var taggedCourses = Courses.find({"keywords": tag}).fetch();
```
Comment (brylie) [inline client/templates/courses/tagged-courses.js:40]: I added a "keywords" filter to the query, so that it will only return courses where keywords contains tags. This ensures that the page will show only tagged courses, despite other active subscriptions.

Comment (sebbel) [inline client/templates/courses/tagged-courses.js:40]: Thanks for your Work, @brylie! We should add another issue for fixing the
subscriptions on a global scale. I am currently not at the computer but
will check later

Am Mittwoch, 27. Januar 2016 schrieb Brylie Christopher Oxley :

**Comment by brylie** [comment]
Yep. Lets basically deprecate Iron Router, and migrate to Flow Router. This will include moving subscriptions to templates, removing controller code, and rewriting our routes.

**Comment by sebbel** [comment]
What are the benefits of flow router? I have not worked with it extensively

Am Mittwoch, 27. Januar 2016 schrieb Brylie Christopher Oxley :

**Comment by brylie** [comment]
The [Meteor Guide recommends Flow Router](http://guide.meteor.com/routing.html#flow-router) as the routing solution.

For more about the differences, see the following:
- [The State of Routing in Meteor](https://themeteorchef.com/blog/the-state-of-routing-in-meteor/)
- [FlowRouter or IronRouter](https://forums.meteor.com/t/flowrouter-or-ironrouter/4705) discussion on Meteor forums
- [FlowRouter 2.0 & the Meteor Routing Guide](https://forums.meteor.com/t/flowrouter-2-0-the-meteor-routing-guide/7699) announcement on Meteor forums

**Comment by Rahbaran** [comment]
Works now. Great. One little bug, though:

You can't jump from let's say `http://localhost:3000/learn/insects` to `http://localhost:3000/learn/bees`. The template won't re-render. It's necessary to leave the `/:tag` route completely.

**Comment by sebbel** [comment]
Problem here seems to be that onCreated in taggedCourses is only run once, since it's using the same template. I found this issue https://github.com/iron-meteor/iron-router/issues/1337
but can't apply what I've read there to our project. Somehow it does not work.

**Comment by brylie** [comment]
Ok. We need to reactively fetch the URL parameter andsubscribe to the tagged courses from within an `instance.autorun()` function. The autorun can be defined in `onCreated` callback. This way, when the URL parameter changes, the subscription code is re-run - returning the relevant courses.

**Comment by sebbel** [comment]
tried that yesterday but somehow the helpers stop being reactive

Am Donnerstag, 28. Januar 2016 schrieb Brylie Christopher Oxley :

**Comment by brylie** [comment]
The [Meteor Routing Guide](https://kadira.io/academy/meteor-routing-guide/) provides an [example of reactive subscription management in Blaze templates using Flow Router](https://kadira.io/academy/meteor-routing-guide/content/subscriptions-and-data-management/with-blaze)

**Comment by brylie** [comment]
Helpers are reactive contexts. What helper do you mean? What reactive data source changes within the helper function, to trigger it to re-run?

**Comment by brylie** [comment]
Check out the [reactivity section of the Meteor docs](http://docs.meteor.com/#/full/reactivity) for a bit more clarification.

**Comment by brylie** [comment]
One possibility is to wrap the helper code in the following:

``` js
let instsnce = Template.instance();

if (instance.subscriptionsReady()) {
  // Fetch tagged courses ...
}
```

The helper code should be re-run when the subscription autorun completes, since DB cursor `ready()` is a resctive data source.

**Comment by sebbel** [comment]
:kiss: @brylie. that was just what was missing. Tag functionality seems to work now,
@brylie or @Rahbaran please review.

**Comment by Rahbaran** [comment]
Will do tonight.

**Comment by Rahbaran** [comment]
Just reviewed it. Works like a charm. @brylie: please merge if you're happy with the code.
